File: lib/services/firestore_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/medicine_model.dart';
import '../models/dose_record_model.dart';

class FirestoreService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  CollectionReference<Map<String, dynamic>> _medicinesRef(String userId) {
    return _firestore.collection('users').doc(userId).collection('medicines');
  }

  CollectionReference<Map<String, dynamic>> _doseRecordsRef(String userId) {
    return _firestore.collection('users').doc(userId).collection('doseRecords');
  }

  // ── Medicines ──────────────────────────────────────────────

  Stream<List<Medicine>> getMedicinesStream(String userId) {
    return _medicinesRef(userId)
        .orderBy('createdAt', descending: false)
        .snapshots()
        .map((snapshot) {
      final List<Medicine> medicines = [];
      for (final doc in snapshot.docs) {
        medicines.add(Medicine.fromMap(doc.id, doc.data()));
      }
      return medicines;
    });
  }

  Future<void> addMedicine(String userId, Medicine medicine) async {
    await _medicinesRef(userId).add(medicine.toMap());
  }

  Future<void> updateMedicine(String userId, Medicine medicine) async {
    await _medicinesRef(userId).doc(medicine.id).update(medicine.toMap());
  }

  Future<void> deleteMedicine(String userId, String medicineId) async {
    await _medicinesRef(userId).doc(medicineId).delete();
  }

  Future<void> decrementAmount(String userId, String medicineId, int currentAmount) async {
    if (currentAmount > 0) {
      await _medicinesRef(userId).doc(medicineId).update({
        'amount': currentAmount - 1,
      });
    }
  }

  // ── Dose Records ───────────────────────────────────────────

  Future<void> recordDose(String userId, DoseRecord record) async {
    await _doseRecordsRef(userId).add(record.toMap());
  }

  Future<List<DoseRecord>> getDoseRecordsForDate(
      String userId, String date) async {
    final QuerySnapshot<Map<String, dynamic>> snapshot =
        await _doseRecordsRef(userId)
            .where('scheduledDate', isEqualTo: date)
            .get();

    final List<DoseRecord> records = [];
    for (final doc in snapshot.docs) {
      records.add(DoseRecord.fromMap(doc.id, doc.data()));
    }
    return records;
  }
}

File: MediMinderSwiftUI/MediMinderSwiftUI/Services/FirestoreService.swift
import Foundation
import FirebaseFirestore
import Combine

/// Wraps all Firestore read/write operations.
/// Mirrors the Flutter firestore_service.dart exactly.
class FirestoreService {
    private let db = Firestore.firestore()

    // MARK: - Collection References

    private func medicinesRef(userId: String) -> CollectionReference {
        db.collection("users").document(userId).collection("medicines")
    }

    private func doseRecordsRef(userId: String) -> CollectionReference {
        db.collection("users").document(userId).collection("doseRecords")
    }

    // MARK: - Medicines

    /// Returns a Combine publisher that streams real-time medicine updates.
    /// The Firestore listener is removed automatically when the publisher is cancelled.
    func getMedicinesStream(userId: String) -> AnyPublisher<[Medicine], Error> {
        let subject = PassthroughSubject<[Medicine], Error>()

        let listener = medicinesRef(userId: userId)
            .order(by: "createdAt", descending: false)
            .addSnapshotListener { snapshot, error in
                if let error = error {
                    subject.send(completion: .failure(error))
                    return
                }
                let medicines = snapshot?.documents.compactMap { Medicine(document: $0) } ?? []
                subject.send(medicines)
            }

        return subject
            .handleEvents(receiveCancel: { listener.remove() })
            .eraseToAnyPublisher()
    }

    func addMedicine(userId: String, medicine: Medicine) async throws {
        try await medicinesRef(userId: userId).addDocument(data: medicine.toFirestoreData())
    }

    func updateMedicine(userId: String, medicine: Medicine) async throws {
        try await medicinesRef(userId: userId)
            .document(medicine.id)
            .setData(medicine.toFirestoreData())
    }

    func deleteMedicine(userId: String, medicineId: String) async throws {
        try await medicinesRef(userId: userId).document(medicineId).delete()
    }

    func decrementAmount(userId: String, medicineId: String, currentAmount: Int) async throws {
        guard currentAmount > 0 else { return }
        try await medicinesRef(userId: userId).document(medicineId).updateData([
            "amount": currentAmount - 1
        ])
    }

    // MARK: - Dose Records

    func recordDose(userId: String, record: DoseRecord) async throws {
        try await doseRecordsRef(userId: userId).addDocument(data: record.toFirestoreData())
    }

    func getDoseRecordsForDate(userId: String, date: String) async throws -> [DoseRecord] {
        let snapshot = try await doseRecordsRef(userId: userId)
            .whereField("scheduledDate", isEqualTo: date)
            .getDocuments()
        return snapshot.documents.compactMap { DoseRecord(document: $0) }
    }
}
